### PR TITLE
Add fallback date formats to bs-datepicker

### PR DESCRIPTION
**Pull request by austin-bittinger**
Allows the addition of an array of fallback date formats that can be used to parse the typed input

### CODE REVIEW

**Comment by VictorioBerra** [comment]
lgtm

File: src/datepicker/datepicker.js [hunk 1]
```diff
@@ -346,7 +349,9 @@ angular.module('mgcrea.ngStrap.datepicker',
 					'onBeforeShow',
 					'onShow',
 					'onBeforeHide',
-					'onHide'
+					'onHide',
+					'onInvalid',
```
Comment (austin-bittinger) [inline src/datepicker/datepicker.js]: These are the callbacks, following the angularstrap convention of callbacks being prefixed with ```bs-```

File: src/datepicker/datepicker.js [hunk 2]
```diff
@@ -454,8 +459,11 @@ angular.module('mgcrea.ngStrap.datepicker',
 					controller.$setValidity('date', isValid);
 					controller.$setValidity('min', isMinValid);
 					controller.$setValidity('max', isMaxValid);
+
 					// Only update the model when we have a valid date
 					if (isValid) controller.$dateValue = parsedDate;
+
+					return isValid && isMaxValid && isMinValid;
```
Comment (austin-bittinger) [inline src/datepicker/datepicker.js]: Changed this to return whether everything was valid so we can save the amount of calls to ```triggerInvalid()```

File: src/datepicker/datepicker.js [hunk 3]
```diff
@@ -472,16 +480,29 @@ angular.module('mgcrea.ngStrap.datepicker',
 					return;
 				}
 
+				function triggerValid() {
```
Comment (austin-bittinger) [inline src/datepicker/datepicker.js]: These two helper methods simply check if the callback was actually defined, and if it is, call it

File: dist/modules/datepicker.js [hunk 4]
```diff
@@ -271,6 +272,15 @@ angular.module('mgcrea.ngStrap.datepicker', [ 'mgcrea.ngStrap.helpers.dateParser
         lang: lang,
         strict: options.strictFormat
       });
+      var fallbackParsers = [];
+      for (var i = 0; i < options.fallbackFormats.length; i++) {
```
Comment (cdumas22) [inline dist/modules/datepicker.js]: use angular.forEach

File: dist/modules/datepicker.js [hunk 5]
```diff
@@ -322,6 +332,17 @@ angular.module('mgcrea.ngStrap.datepicker', [ 'mgcrea.ngStrap.helpers.dateParser
         controller.$setValidity('max', isMaxValid);
         if (isValid) controller.$dateValue = parsedDate;
       }
+      function tryFallbackFormats(viewValue) {
+        var result;
+        for (var i = 0; i < fallbackParsers.length; i++) {
```
Comment (cdumas22) [inline dist/modules/datepicker.js]: use angular.forEach

**Review by cdumas22 — DISMISSED** [review]

**Comment by austin-bittinger** [comment]
@cdumas22 Latest fixes are up

File: tasks/test.js [hunk 6]
```diff
@@ -6,84 +6,80 @@ var Server = require('karma').Server;
 var reporter = require('./helpers/reporter');
```
Comment (cdumas22) [inline tasks/test.js]: fix the whitespace to match their standard

**Comment by austin-bittinger** [comment]
@cdumas22 I went through and restored the old versions of any files I changed that didn't need to be touched for the PR

**Pull request merged**